packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,89 @@
+/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0 */
+import { IRole, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
+import { IBucket } from "aws-cdk-lib/aws-s3";
+import {
+  ApiGatewayIntegration,
+  Integration,
+  IntegrationRenderProps,
+} from "./integration";
+import { generateCorsResponseParameters } from "../prepare-spec-event-handler/prepare-spec";
+import { bucketInvocationUri } from "../spec/utils";
+
+export interface S3IntegrationProps {
+  /**
+   * The S3 bucket to be invoked on integration
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * The IAM role to be used to grant permissions to the API Gateway to invoke the S3 bucket
+   * @default - a new role will be created
+   */
+  readonly role?: IRole;
+
+  /**
+   * The path override for the integration
+   * @default - integration path is used
+   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html
+   */
+  readonly pathOverride?: string;
+}
+
+/**
+ * An S3 integration
+ */
+export class S3Integration extends Integration {
+  private readonly bucket: IBucket;
+  private readonly role: IRole;
+  private readonly pathOverride?: string;
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+
+    if (props.role) {
+      this.role = props.role;
+    } else {
+      this.role = new Role(this.bucket, "Role", {
+        assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
+      });
+    }
+
+    this.bucket.grantReadWrite(this.role, this.pathOverride ?? "*");
+    this.pathOverride = props.pathOverride;
+  }
+
+  /**
+   * Render the lambda integration as a snippet of OpenAPI
+   */
+  public render(props: IntegrationRenderProps): ApiGatewayIntegration {
+    return {
+      type: "AWS",
+      httpMethod: props.method.toUpperCase(),
+      uri: bucketInvocationUri(
+        props.scope,
+        this.bucket,
+        this.pathOverride ?? props.path
+      ),
+      credentials: this.role.roleArn,
+      responses: {
+        default: {
+          statusCode: "200",
+          responseParameters: props.corsOptions
+            ? generateCorsResponseParameters(props.corsOptions)
+            : {},
+          responseTemplates: {},
+        },
+        "4|5\\d{2}": {
+          statusCode: "500", // force 500 because 4XX errors is not a real client error, it's a misconfiguration on the server side
+          responseParameters: props.corsOptions
+            ? generateCorsResponseParameters(props.corsOptions)
+            : {},
+          responseTemplates: {},
+        },
+      },
```

Comment:
Ooh sorry I think one more thing we should probably support is path parameters, like in that example where the path parameters can be used in the s3 key from the docs: https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html#api-items-in-folder-as-s3-objects-in-bucket

I think customising the bucket is definitely out of scope, but perhaps we just support a simple mapping where the path parameters are all included in the integration path parameters and so used for variable keys 😄 

I think to achieve this all you'd need to do is add the same thing as I did for the custom nlb integration to `requestParameters`:

https://github.com/aws/aws-pdk/blob/917a5db21d96d620ade7f306f9fedbca16ad1648/packages/type-safe-api/docs/developer_guides/type-safe-api/custom_integration_ecs.md?plain=1#L242-L249

I think we'd also get the ability to use them differently in the overridden path for free which is also a pretty cool feature! :) eg the api might be `/pets/{petId}` which might map to `<bucket>/pet-data/{petId}/details.json` or something :)

Reply:
I don't understand how it could map something like in the tutorial

```
{bucket}: method.request.path.folder
```

with:
- {bucket}: the path parameter override
- folder: the real path parameter

Reply:
So I think we want arbitrary path parameters to be mapped as defined in the Smithy/OpenAPI model :) Just mapping them to the same name like `{bucket}` to `{bucket}` is simplest rather than renaming like in the tutorial :)

So I think as long as we add the above mapping it'd map them? Worth testing though! :)

Reply:
Ok your're right, I think that the tutorial overcomplicated the thing with different names too